File: src/hotspot/share/prims/nativeMemoryTracking.cpp
/*
 * Copyright (c) 2026, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 *
 */

#include "jni.h"
#include "nmt/mallocHeader.hpp"
#include "nmt/mallocHeader.inline.hpp"
#include "nmt/memTagFactory.hpp"
#include "nmt/memTracker.hpp"
#include "runtime/interfaceSupport.inline.hpp"
#include "runtime/jniHandles.hpp"
#include "runtime/jniHandles.inline.hpp"

#include <limits>

#define CC (char*)  /*cast a literal from (const char*)*/
#define FN_PTR(f) CAST_FROM_FN_PTR(void*, &f)

JVM_ENTRY(jlong, NMT_makeTag(JNIEnv *env, jobject ignored_this, jobject tag_name_string)) {
  constexpr static int max_tagname_len = 1024;
  if (!MemTracker::enabled()) {
    return (jlong)mtNone;
  }
  Handle tnh(Thread::current(), JNIHandles::resolve(tag_name_string));
  oop oop = tnh();

  // The tag name must be a non-null instance of java.lang.String
  // whose length doesn't exceed max_tagname_len.
  if (oop == nullptr) {
    return -1;
  }
  if (!oop->is_a(vmClasses::String_klass())) {
    return -1;
  }
  typeArrayOop value = java_lang_String::value(oop);
  int length = java_lang_String::length(oop, value);
  if (length > max_tagname_len) {
    return -1;
  }

  // OK, fine.
  ResourceMark rm;
  const char* tag_name = java_lang_String::as_utf8_string(oop);
  MemTag tag = MemTagFactory::tag(tag_name);
  return (jlong)tag;
}
JVM_END

JVM_ENTRY(jlong, NMT_allocate0(JNIEnv *env, jobject ignored_this, jlong size, jlong mem_tag)) {
  if (size < 0) {
    return 0;
  }
  bool enabled = MemTracker::enabled();
  if (enabled && (mem_tag < 0 || MemTagFactory::number_of_tags() <= mem_tag)) {
    return 0;
  }

#ifndef _LP64
  assert(sizeof(jlong) > sizeof(size_t), "must be");
  if (size > static_cast<jlong>(std::numeric_limits<size_t>::max())) {
    return 0;
  }
#endif

  MemTag tag = enabled ? (MemTag)mem_tag : mtOther;
  return (jlong)os::malloc(size, tag);
}
JVM_END

JVM_ENTRY(jboolean, NMT_isNMTEnabled(JNIEnv *env, jobject ignored_this)) {
  return (jboolean)MemTracker::enabled();
}
JVM_END

static JNINativeMethod NMT_methods[] = {
  {CC "makeTag", CC "(Ljava/lang/String;)J", FN_PTR(NMT_makeTag)},
  {CC "allocate0", CC "(JJ)J", FN_PTR(NMT_allocate0)},
  {CC "isNMTEnabled", CC "()Z", FN_PTR(NMT_isNMTEnabled)}
};

JVM_ENTRY(void, JVM_RegisterNativeMemoryTrackingMethods(JNIEnv *env, jclass NMT_class)) {
  ThreadToNativeFromVM ttnfv(thread);

  int status = env->RegisterNatives(NMT_class, NMT_methods, sizeof(NMT_methods) / sizeof(JNINativeMethod));
  guarantee(status == JNI_OK && !env->ExceptionCheck(),
            "register java.lang.invoke.MethodHandleNative natives");
}
JVM_END

File: src/java.base/share/classes/jdk/internal/foreign/NativeMemoryTracking.java
/*
 * Copyright (c) 2026, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.internal.foreign;

import java.lang.OutOfMemoryError;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.MemorySegment.Scope;

import jdk.internal.access.JavaLangAccess;
import jdk.internal.access.SharedSecrets;
import jdk.internal.foreign.ArenaImpl;
import jdk.internal.foreign.MemorySessionImpl;
import jdk.internal.foreign.NativeMemorySegmentImpl;
import jdk.internal.misc.Unsafe;

/**
 * The entrypoint to the static NMT interface.
 */
public class NativeMemoryTracking {
    private static final Unsafe UNSAFE = Unsafe.getUnsafe();
    private static final int addressSize = UNSAFE.addressSize();
    private static final boolean nmtEnabled;

    private static final JavaLangAccess JLACCESS = SharedSecrets.getJavaLangAccess();

    private static native long makeTag(String name);
    private static native long allocate0(long size, long mem_tag);
    private static native boolean isNMTEnabled();
    private static native void registerNatives();


    static {
        runtimeSetup();
        nmtEnabled = isNMTEnabled();
    }

    public static long allocate(long size) throws OutOfMemoryError {
        if (size < 0 ||
            addressSize == 4 && size >>> 32 != 0) {
            throw new IllegalArgumentException();
        }

        // Mimic Unsafe::allocateMemory
        if (size == 0) {
            return 0;
        }

        long ptr = allocate0(size, currentTag());
        if (ptr == 0) {
            throw new OutOfMemoryError();
        }
        return ptr;
    }

    public static final class Tag {
        private long tag;
        Tag(long tag) { this.tag = tag; }
    }

    public static Tag resolveName(String name) {
        if (name == null) {
            throw new IllegalArgumentException("Invalid name, cannot be null");
        }
        if (!nmtEnabled) {
            return new Tag(currentTag());
        }
        long tag = makeTag(name);
        if (tag == -1) {
            throw new IllegalArgumentException();
        }
        return new Tag(tag);
    }

    // Get the current tag as a long
    private static long currentTag() {
        return JLACCESS.nativeMemoryTrackingTag();
    }


    // Replaces the old tag with `newTag` and returns the old one.
    public static Tag getAndSet(Tag newTag) {
        long tag = JLACCESS.nativeMemoryTrackingTag();
        JLACCESS.setNativeMemoryTrackingTag(newTag.tag);
        return new Tag(tag);
    }

    public static void with(Tag tag, Runnable action) {
        Tag oldTag = getAndSet(tag);
        try {
            action.run();
        } finally {
            getAndSet(oldTag);
        }
    }

    private static void runtimeSetup() {
        registerNatives();
    }
}
